Fixes 11334/11354 Multi line code blocks

tonyfast opened the pull request:
This pull request merges the recent input transformer changes from master.  It builds off of the work by @bartskowron in  #11354 to resolve #11334 .  The changes pass tests locally, including the new tests provided by @Carreau; let's see what CI says.

@takluyver [I commented out an assertion](https://github.com/ipython/ipython/blame/498f5f716d102a572cac3571e13426aa002f8454/IPython/core/inputtransformer2.py#L252).  I am not sure what the statement, is this change ok?

bxsx committed:
Fix #11334: Cannot make multi-line code blocks in ipython

When codeop.compile_command() returns None it actually says "at least
some part of the code was compiled successfully" which is not really important
for checking if it's complete or not.
Once we haven't got any errors during compilation process, we just want
to check if there will be another nested block of code or not by checking
a colon.

bxsx committed:
Fix regression...

bxsx committed:
Add a new regression test and fix for it

bxsx committed:
Remove unused assignment

Carreau committed:
add tests

bxsx committed:
always add a trailing newline

tonyfast committed:
Merge branch 'master' into pr/11354

tonyfast committed:
Add some logic to pass all of the check_complete tests

An assert statement needed to be commented out.  I'm not sure what the statement was stating though.

tonyfast commented:
Anyone have any ideas why this may be passing on everything, but Python 3.7dev?

Carreau commented:
> Anyone have any ideas why this may be passing on everything, but Python 3.7dev?

Maybe a change of API in generate_token or similar on Pyhton 3.7 dev.

tonyfast commented:
I'll scope it out. Thanks.

Carreau commented:
that's https://bugs.python.org/issue33899

tonyfast committed:
Use tokenize.NL and tokenize.NEWLINE in the check_complete logic

These changes are necessary for a failure on Python 3.7 dev https://bugs.python.org/issue33899

tonyfast commented:
Thanks for these references @Carreau .  I added some conditions for the newline tokens.  Let's see if this helps.

tonyfast commented:
The CPython changes to tokenize effect 3.6 & 3.7.  The `IPython` tests are not checking on Python3.6dev. 
 Should there be tests for 3.6dev because of this use case?

Carreau commented:
Likely, yes. That would be good. unless it make the tests take too long in
High case we can test only on one -dev

On Thu, Oct 4, 2018, 06:29 Tony Fast <notifications@github.com> wrote:

> The CPython changes to tokenize effect 3.6 & 3.7. The IPython tests are
> not checking on Python3.6dev.
> Should there be tests for 3.6dev because of this use case?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipython/ipython/pull/11358#issuecomment-427018524>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUez7XeLldVbRK8NZ01tJKXwkJuH2X1ks5uhg1IgaJpZM4XHPxz>
> .
>

tonyfast committed:
Add 3.6-dev to the test.

tonyfast commented:
If we consider a simple statement like `a=999`.

In Python 3.6 with the old tokenize:

```
Version info sys.version_info(major=3, minor=6, micro=6, releaselevel='final', serial=0)
[TokenInfo(type=1 (NAME), string='a', start=(1, 0), end=(1, 1), line='a=999'),
 TokenInfo(type=53 (OP), string='=', start=(1, 1), end=(1, 2), line='a=999'),
 TokenInfo(type=2 (NUMBER), string='999', start=(1, 2), end=(1, 5), line='a=999'),
 TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')]
```
compared to in Python 3.7 on release candidate.

```
Version info sys.version_info(major=3, minor=7, micro=1, releaselevel='candidate', serial=1)
[TokenInfo(type=1 (NAME), string='a', start=(1, 0), end=(1, 1), line='a=999'),
 TokenInfo(type=53 (OP), string='=', start=(1, 1), end=(1, 2), line='a=999'),
 TokenInfo(type=2 (NUMBER), string='999', start=(1, 2), end=(1, 5), line='a=999'),
 TokenInfo(type=4 (NEWLINE), string='', start=(1, 5), end=(1, 6), line=''),
 TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')]
```

tonyfast committed:
A refactor to check_complete to pass the test cases.

tonyfast committed:
Fix check_complete with a more verbose approach.

tonyfast commented:
@Carreau it looks like the core tests are passing, but I am receiving a timeout error in the terminal test group .  Do you have an insight into this?

tonyfast committed:
Uncomment the ! assertion.

Co-Authored-By: Nicholas Bollweg <bollwyvl@users.noreply.github.com>

tonyfast commented:
The errors at the moment are in `SystemAssign`.  The test case `=!` is trying to assign a system statement.

tonyfast committed:
Add an extra condition to SystemAssign.

Co-Authored-By: Nicholas Bollweg <bollwyvl@users.noreply.github.com>

tonyfast committed:
Add check_complete test for exit.

Co-Authored-By: Nicholas Bollweg <bollwyvl@users.noreply.github.com>

tonyfast committed:
Remove a condition from check_complete

Co-Authored-By: Nicholas Bollweg <bollwyvl@users.noreply.github.com>

tonyfast commented:
Woo!  The new and old tests are passing now.  These fixes resolve #11334 #11354 #11346 #11337 .

@Carreau @takluyver could you review this when you get a chance?

Carreau commented:
Thanks ! Awesome work @tonyfast !

I'll try to review and merced this week-end.

Carreau commented on .travis.yml:
```diff
@@ -34,6 +34,7 @@ after_success:
 
 matrix:
     include:
+        - { python: "3.6-dev", dist: xenial, sudo: true }
```
I'm just going to add some "ok", to follow where I am in the review for myself.

Carreau replied:
ok.

Carreau commented on IPython/core/inputtransformer2.py:
```diff
@@ -234,6 +234,7 @@ def find(cls, tokens_by_line):
         for line in tokens_by_line:
             assign_ix = _find_assign_op(line)
             if (assign_ix is not None) \
+                    and not line[assign_ix].line.strip().startswith('=') \
```
ok

Carreau commented on IPython/core/inputtransformer2.py:
```diff
@@ -369,11 +370,15 @@ def transform(self, lines):
         line = assemble_continued_line(lines, (start_line, start_col), end_line)
 
-        if line[:2] in ESCAPE_DOUBLES:
+        if len(line) > 1 and line[:2] in ESCAPE_DOUBLES:
```
ok

Carreau commented on IPython/core/inputtransformer2.py:
```diff
@@ -369,11 +370,15 @@ def transform(self, lines):
+        if escape in tr:
+            call = tr[escape](content)
+        else:
+            call = ''
```
ok

Carreau commented on IPython/core/inputtransformer2.py:
```diff
@@ -575,10 +583,28 @@ def check_complete(self, cell: str):
+            # Append an newline for consistent tokenization
+            # See https://bugs.python.org/issue33899
+            cell += '\n'
+
```
Seem good. Looks like we should either ends with 2 \n or None.

Carreau commented on IPython/core/inputtransformer2.py:
```diff
@@ -575,10 +583,28 @@ def check_complete(self, cell: str):
+        if not lines:
+            return 'complete', None
+
+        if lines[-1].endswith('\\'):
```
ok

bartskowron commented on IPython/core/tests/test_inputtransformer2.py:
```diff
@@ -212,6 +214,17 @@ def test_check_complete():
     nt.assert_equal(cc("def a():\n x=1\n global x"), ('invalid', None))
     nt.assert_equal(cc("a \\ "), ('invalid', None))  # Nothing allowed after backslash
+    nt.assert_equal(cc("1\\\n+2"), ('complete', None))
+    nt.assert_equal(cc("1\\\n+2"), ('complete', None))
```
redundant test

tonyfast committed:
Remove redundant test

Nice catch @bartskowron

Carreau commented:
Thanks all, let's try to get that in and have an IPython 7.1 by end of week.
That should match the imminent release of Python 3.7.1 that have the compatibility issues mentioned earlier.
